src/main/java/com/google/sps/servlets/PlanMailServlet.java

```diff
@@ -0,0 +1,217 @@
+import com.google.sps.model.GmailClientFactory;
+import com.google.sps.model.GmailClientImpl;
+import com.google.sps.utility.DatePair;
+import com.google.sps.utility.FreeTimeUtility;
+import com.google.sps.utility.JsonUtility;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.List;
+import java.util.StringTokenizer;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+import javax.mail.MessagingException;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/** GET function responds JSON string containing potential events to read mail. */
+@WebServlet("/plan-mail")
+public class PlanMailServlet extends AuthenticatedHttpServlet {
+  private final CalendarClientFactory calendarClientFactory;
+  private final GmailClientFactory gmailClientFactory;
+
+  /** Create servlet with default CalendarClient and Authentication Verifier implementations */
+  public PlanMailServlet() {
+    calendarClientFactory = new CalendarClientImpl.Factory();
+    gmailClientFactory = new GmailClientImpl.Factory();
+  }
+
+  /**
+   * Create servlet with explicit implementations of CalendarClient and AuthenticationVerifier
+   *
+   * @param authenticationVerifier implementation of AuthenticationVerifier
+   * @param calendarClientFactory implementation of CalendarClientFactory
+   */
+  public PlanMailServlet(
+      AuthenticationVerifier authenticationVerifier,
+      CalendarClientFactory calendarClientFactory,
+      GmailClientFactory gmailClientFactory) {
+    super(authenticationVerifier);
+    this.calendarClientFactory = calendarClientFactory;
+    this.gmailClientFactory = gmailClientFactory;
+  }
+
+  /**
+   * Returns string containing the some of the user's free time intervals, where events can be
+   * created
+   *
+   * @param request Http request from the client. Should contain idToken and accessToken
+   * @param response 403 if user is not authenticated, or Json string with the user's events
+   * @throws IOException if an issue arises while processing the request
+   */
+  @Override
+  public void doGet(
+      HttpServletRequest request, HttpServletResponse response, Credential googleCredential)
+      throws IOException {
+    assert googleCredential != null
+        : "Null credentials (i.e. unauthenticated requests) should already be handled";
+
+    CalendarClient calendarClient = calendarClientFactory.getCalendarClient(googleCredential);
+    GmailClient gmailClient = gmailClientFactory.getGmailClient(googleCredential);
+    long fiveDaysInMillis = TimeUnit.DAYS.toMillis(5);
+    Date timeMin = calendarClient.getCurrentTime();
+    Date timeMax = new Date(timeMin.getTime() + fiveDaysInMillis);
+    List<Event> calendarEvents = getEvents(calendarClient, timeMin, timeMax);
+
+    FreeTimeUtility freeTimeUtility = new FreeTimeUtility(timeMin);
+    for (Event event : calendarEvents) {
```

Comment:
This should probably be offloaded into a helper method for readability.

Reply:
I can probably do it for now, as it is only updating the `freeTimeUtility`, but eventually I will need to get the `preAssigned` time variable (of type long) from the `for` loop. Is there a way you know I could do that as well?